Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -2212,19 +2230,143 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+
     if (commandName == "scrollToEndOfDocument:"_s)
-        return keyboardAnimator->beginKeyboardScrollGesture(ScrollDirection::ScrollDown, ScrollGranularity::Document, false);
+        return checkedKeyboardScrollingAnimator()->beginKeyboardScrollGesture(ScrollDirection::ScrollDown, ScrollGranularity::Document, false);
+
+    return false;
+}
+
+bool UnifiedPDFPlugin::handleKeyboardEventForDiscreteDisplayMode(const WebKeyboardEvent& event)
+{
+    if (!m_currentlySnappedPage || !isInDiscreteDisplayMode())
+        return false;
+
+    if (event.type() == WebEventType::KeyUp) {
+        checkedKeyboardScrollingAnimator()->handleKeyUpEvent();
+        return false;
+    }
+
+    if (event.type() != WebEventType::KeyDown || m_isScrollingWithAnimationToPageExtent)
+        return false;
+
+    auto key = event.key();
+    if (key == "ArrowLeft"_s) {
+        if (checkedKeyboardScrollingAnimator()->beginKeyboardScrollGesture(ScrollDirection::ScrollLeft, ScrollGranularity::Line, false)) {
+            m_animatedKeyboardScrollingDirection = ScrollDirection::ScrollLeft;
+            return true;
+        }
+        return false;
+    }
+
+    if (key == "ArrowRight"_s) {
+        if (checkedKeyboardScrollingAnimator()->beginKeyboardScrollGesture(ScrollDirection::ScrollRight, ScrollGranularity::Line, false)) {
+            m_animatedKeyboardScrollingDirection = ScrollDirection::ScrollRight;
+            return true;
+        }
+        return false;
+    }
+
+    auto currentPage = *m_currentlySnappedPage;
+    auto currentPageContentRect = pageBoundsInContentsSpace(currentPage);
+    bool isTallerThanViewport = currentPageContentRect.height() > m_size.height();
+    if (isTallerThanViewport) {
+        if (key == "ArrowUp"_s) {
+            if (snapToNearbyPageExtentForKeyboardScrolling(ScrollDirection::ScrollUp))
+                return true;
+
+            if (checkedKeyboardScrollingAnimator()->beginKeyboardScrollGesture(ScrollDirection::ScrollUp, ScrollGranularity::Line, false)) {
+                m_animatedKeyboardScrollingDirection = ScrollDirection::ScrollUp;
+                return true;
+            }
+        }
+
+        if (key == "ArrowDown"_s) {
+            if (snapToNearbyPageExtentForKeyboardScrolling(ScrollDirection::ScrollDown))
+                return true;
+
+            if (checkedKeyboardScrollingAnimator()->beginKeyboardScrollGesture(ScrollDirection::ScrollDown, ScrollGranularity::Line, false)) {
+                m_animatedKeyboardScrollingDirection = ScrollDirection::ScrollDown;
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    if (key == "ArrowUp"_s && currentPage) {
+        auto previousPageContentRect = pageBoundsInContentsSpace(currentPage - 1);
+        return scrollAnimator().scrollToPositionWithoutAnimation({ static_cast<float>(m_scrollOffset.width()), previousPageContentRect.y() });
+    }
```

Comment:
Do we not want the same kind of "skip two pages if necessary" behavior? As you did for ArrowDown

Reply:
So this is only necessary when moving down, since the current page index in 2-up mode will be `0`, `2`, `4`, `6`, … (so going back by 1 page will yield the correct previous offset)